ipa-core/src/protocol/basics/check_zero.rs

```diff
@@ -124,4 +187,122 @@ mod tests {
 
         Ok(())
     }
+
+    pub struct NotifyOnceCell<T: Clone + Send + Sync> {
+        inner: Mutex<NotifyOnceCellInner<T>>,
+    }
+
+    struct NotifyOnceCellInner<T: Clone + Send + Sync> {
+        cell: OnceCell<T>,
+        notify: Arc<Notify>,
+    }
+
+    impl<T: Clone + Send + Sync> NotifyOnceCell<T> {
+        pub fn new() -> Self {
+            Self {
+                inner: Mutex::new(NotifyOnceCellInner {
+                    cell: OnceCell::new(),
+                    notify: Arc::new(Notify::new()),
+                }),
+            }
+        }
+
+        pub fn set(&self, value: T) -> Result<(), T> {
+            let inner = self.inner.lock().unwrap();
+            inner.cell.set(value)?;
+            inner.notify.notify_waiters();
+            Ok(())
+        }
+
+        pub fn get(&self) -> Pin<Box<dyn Future<Output = T> + Send + '_>> {
+            let inner = self.inner.lock().unwrap();
+            if let Some(value) = inner.cell.get() {
+                return ready(value.clone()).boxed();
+            }
+            let notify = inner.notify.clone();
+            async move {
+                notify.notified().await;
+                self.inner.lock().unwrap().cell.get().unwrap().clone()
+            }
+            .boxed()
+        }
+    }
+
+    struct MaliciousCheckZeroInterceptor {
```

Comment:
lets explain what is the goal for this interceptor because it does seem very specific to check zero use case